### ISSUE TITLE
callback / cannot yield intermediate images on the fly during inference

### ISSUE THREAD

**Issue by Clement-Lelievre** [huggingface/diffusers#9407]
Hi, 

in advance apologies if this has been asked already, or if I'm just misusing the diffusers API.

Using `diffusers==0.30.2`

**What API design would you like to have changed or added to the library? Why?**

I will illustrate straight away the general issue with my use case: I need to call a (FLUX) diffusers pipeline from some endpoint of mine, passing a callback that decodes latents and saves on disk intermediate images obtained from them, at the end of each step. So far, so good: I do manage to get the intermediate images saved on disk. I do this using the pipeline argument `callback_on_step_end`

Now, I'd like to _**yield**_ (in the pythonic meaning) these intermediate images on the fly, as soon as they're available, ie at the end of each inference step. I need to do so from my endpoint. That's where my problem is.

I could not make this idea work using with diffusers callback mechanism.
I mean, I did manage that by subclassing the pipeline, copy-pasting the dunder call method code and overriding it, but this is not maintainable, especially since the FLUX code evolves rapidly nowadays.
Also, note that currently diffusers assigns the result of the call to the callback to a variable and expects it to implement the `.pop` method, which might add constraints (diffusers typically expects a kwarg dict, see [here](https://github.com/huggingface/diffusers/blob/v0.30.2/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L1026)).

Another approach I thought of is to monitor the disk contents in a parallel process during the call to the pipeline.

But is there an easier way?

**What use case would this enable or better enable? Can you give us a code example?**

This allows to manipulate the objects produced by the callback live, instead of having to wait for the whole reverse diffusion to finish.

Thank you

cc @sayakpaul @yiyixuxu

also tagging @asomoza since I saw he is the contributor to the official callback interface

**Comment by asomoza** [comment]
Since I come from other languages I find that `yield` is something really restrictive when you use async code or threads, also you're saying you need the images as soon as they're available and with the callbacks you get that too, is not that the callback only gets called after the whole generation is done but on each step too.

This is more of an issue on your side or the UI side than on the diffusers API, for example, I did a demo with Gradio and to display the images in an easy way you can use `yield`, so I had to do the same as you're suggesting, to change the pipeline because I can't (without manually using websockets) update the UI from the callback function.

But, if you use a more advanced system that lets you manually manage the state of the UI like PyQT, you can use the callbacks without a problem since you can update the UI as soon as you get the images from the callback.

In short what you really need is to find a way to communicate with your UI from the callback on each step, there isn't a method to do it with `yield` at the moment. 

cc: @yiyixuxu in case this is something planned in the future or not.

**Comment by Clement-Lelievre** [comment]
Hi @asomoza , thanks for your reply, let me clarify or restate some points:

>  the callback only gets called after the whole generation is done but on each step too.

yes, I do know that the callback gets executed at the end of each step, hence the name `on_step_end`

> also you're saying you need the images as soon as they're available and with the callbacks you get that too

I do get the images saved on disk (hence my idea of minotiring the disk contents in a parallel process), but I can't access them in code, or at least I don't know how to do this

> In short what you really need is to find a way to communicate with your UI from the callback on each step

I have no UI, here I am working from an API endpoint. No UI involved. My use case is more like you're calling a diffusers pipeline in python and you'd like to access the intermediate images in code, while the inference happens.

**Comment by asomoza** [comment]
My apologies, I think I misunderstood and over simplified your problem. 

I still don't understand your problem though and this is probably because, as I stated, I don't really understand and use yield unless I'm forced, because as I wrote, I'm used to do it manually by myself.

Probably better if someone else tries to help you, just as a clarification, what I don' understand in your problem is this:

> ... I'd like to yield (in the pythonic meaning) these intermediate images on the fly, as soon as they're available, ie at the end of each inference step

but then you wrote:

> I do get the images saved on disk...., but I can't access them in code...

If you're saving them, you can access them in code, so that's why I though you need them for a specific UI which needs the images with `yield` or something similar. 

In short I don't see why instead of saving them to disk you use them to whatever you need in your endpoint. To me this seems like the problem people have in javascript or node with async/await programing and I associated them by mistake....maybe.

**Comment by Clement-Lelievre** [comment]
> My apologies, I think I misunderstood and over simplified your problem.
> 
> I still don't understand your problem though and this is probably because, as I stated, I don't really understand and use yield unless I'm forced, because as I wrote, I'm used to do it manually by myself.
> 
> Probably better if someone else tries to help you, just as a clarification, what I don' understand in your problem is this:
> 
> > ... I'd like to yield (in the pythonic meaning) these intermediate images on the fly, as soon as they're available, ie at the end of each inference step
> 
> but then you wrote:
> 
> > I do get the images saved on disk...., but I can't access them in code...
> 
> If you're saving them, you can access them in code, so that's why I though you need them for a specific UI which needs the images with `yield` or something similar.
> 
> In short I don't see why instead of saving them to disk you use them to whatever you need in your endpoint. To me this seems like the problem people have in javascript or node with async/await programing and I associated them by mistake....maybe.


To clarify, I'd like to call the pipeline in order to make an inference, AND to access the intermediate images in code from memory at the same time, as soon as they're produced.
to illustrate my need, in pure python this could be something like:

```python
for intermediate_image in pipeline(prompt=..., other_args..., callback_on_step_end=decode_latents):
    yield intermediate_image
```

and also somehow get the last, final image

I do not knwo whether this is possible with some async code or via a thread

**Comment by asomoza** [comment]
yeah, that what I answered in the first post, currently this:

```python
for intermediate_image in pipeline
```

Is not possible since for that we need to change the return type of all the pipelines to `yield` and I don't know if that is something planned in the future or not, so lets wait for that answer.

> I do not knwo whether this is possible with some async code or via a thread

This I can assure you that can be done, I do it all the time in a personal app I use for testing diffusers.

**Comment by Clement-Lelievre** [comment]
> I do not knwo whether this is possible with some async code or via a thread
> 
> This I can assure you that can be done, I do it all the time in a personal app I use for testing diffusers.

OK, if you have some material on this explaining how to set it up that's be appreciated!

thks

**Comment by asomoza** [comment]
I don't know your specific use case or what you need to do, but I wanted to do a simple app to test the SD3 preview and probably will expand in the future for masks and inpainting, I did a quick PoC of one with PyQT6 to demonstrate how can you use the callbacks to access the intermediate images.

It's a minimal one file [app](https://github.com/asomoza/diffusers-preview), so it should be easy to understand.

[Here](https://github.com/asomoza/diffusers-preview/blob/a35dc020df32f72a6e398369b9c2acee1e875228/main.py#L68) I set the callback and [here](https://github.com/asomoza/diffusers-preview/blob/a35dc020df32f72a6e398369b9c2acee1e875228/main.py#L125) I manipulate the intermediate images to display them.

PyQT uses "signals" to communicate between processes and threads, if what you're using doesn't have something similar, then it will be a lot more "involved" to implement. 

I think this is outside the scope of diffusers though, since it's a problem on the library/app on top of it on how you can work with the callbacks.

**Comment by Clement-Lelievre** [comment]
update, I run the pipeline call in a thread, pass a callback that saves latents on disk, and in the meantime I watch the disk in the main process and `yield` new images
there are some tricky things to consider, such as avoiding to yield while an image is being written on disk , otherwise the image yielded is partial, but apart from this it works as a charm
